mixedprecision/roots/rpoly.cpp

```diff
@@ -0,0 +1,1052 @@
+template <typename Real>
+Real FindLinearPolynomialRoots(const Real a, const Real b) {
+  return -b / a;
+}
+
+// Stable quadratic roots according to BKP Horn.
+// http://people.csail.mit.edu/bkph/articles/Quadratics.pdf
+template <typename Real>
+void FindQuadraticPolynomialRoots(const Real a, const Real b, const Real c,
+                                  std::vector<std::complex<Real>>& roots) { // TODO: std::complex does not work with arbitrary typess
+  
+    using namespace sw::universal;
+    using std::sqrt, std::abs;
+    const Real D = b * b - 4 * a * c;
+    const Real sqrt_D = sqrt(abs(D));
+
+    // Real roots.
+    if (D >= 0) {
+        if (b >= 0) {
+            roots[0] = std::complex<Real>((-b - sqrt_D) / (2.0 * a), 0);
+            roots[1] = std::complex<Real>((2.0 * c) / (-b - sqrt_D), 0);
+        } else {
+            roots[0] = std::complex<Real>((2.0 * c) / (-b + sqrt_D), 0);
+            roots[1] = std::complex<Real>((-b + sqrt_D) / (2.0 * a), 0);
+        }
+        return;
+    }
+
+    // Use the normal quadratic formula for the complex case.
+    roots[0] = std::complex<Real>(-b / (2.0 * a), sqrt_D / (2.0 * a));
+    roots[1] = std::complex<Real>(-b / (2.0 * a), -sqrt_D / (2.0 * a));
+}
+
+// Perform division by a linear term of the form (z - x) and evaluate P at x.
+template <typename Real>
+void SyntheticDivisionAndEvaluate(const sw::universal::blas::vector<Real>& polynomial,
+                                  const Real x,
+                                  sw::universal::blas::vector<Real>& quotient,
+                                  Real& eval) {
+    SetZero(quotient, polynomial.size() - 1);
+  
+    quotient[0] = polynomial[0];
+    for (size_t i = 1; i < polynomial.size() - 1; i++) {
+        quotient[i] = polynomial[i] + quotient[i - 1] * x;
+    }
+    eval = polynomial[polynomial.size() - 1] + quotient[quotient.size() - 1] * x;
+}
+
+// Perform division of a polynomial by a quadratic factor. The quadratic divisor should have leading 1s.
+template <typename Real>
+void QuadraticSyntheticDivision(const sw::universal::blas::vector<Real>& polynomial,
+                                const sw::universal::blas::vector<Real>& quadratic_divisor,
+                                sw::universal::blas::vector<Real>& quotient,
+                                sw::universal::blas::vector<Real>& remainder) {
+    SetZero(quotient, polynomial.size() - 2);
+    SetZero(remainder, 2);
+
+    quotient[0] = polynomial[0];
+
+    // If the quotient is a constant then polynomial is degree 2 and the math is simple.
+    if (quotient.size() == 1) {
+        // TODO: maybe this could be written cleaner??
+        remainder = {polynomial[1] - polynomial[0] * quadratic_divisor[1], polynomial[2] - polynomial[0] * quadratic_divisor[2]};
+        return; 
+    }
+
+    quotient[1] = polynomial[1] - polynomial[0] * quadratic_divisor[1];
+    for (size_t i = 2; i < polynomial.size() - 2; i++) {
```

Comment:
remember, ++i does not need a temporary, i++ does, so ++i is the preferred from in the loop spec